### ISSUE TITLE
BDR Group Broken When Nodes Unavailable

### ISSUE THREAD

**Issue by srkunze** [2ndQuadrant/bdr#147]
I finally set up my BDR group with 6 LXC containers successfully.

I observed a something odd for an autonomous replication scenario. I removed two "broken" nodes, so I got 4 nodes remaining.

Then, I could neither

1) add a replacement node (which is odd if you imagine this to be a HA setup)
2) create a table (might break some applications)

I could fix this by "announcing the removal of" the previously deleted nodes with `bdr_part_by_node_names`. Then 1) and 2) worked again.

This seems to be bug.

**Comment by ringerc** [comment]
It's not a bug. It's entirely by design.

BDR is intended for deployments where network connectivity may be
intermittent, slow and unreliable.

Under those circumstances, how should it know that you deleted the node?
How long should it wait for it to reconnect before assuming it is
permanently gone? There is no single right answer.

If you delete a node you have to tell BDR it isn't coming back.

The reason you can't join new nodes or do DDL while nodes are unreachable
is that those operations require cluster-wide consensus. If you need to do
those things urgently and you have a node that is down for an extended
period you can part the node, then re-create it later. The downside is, of
course, that you won't be able to replay writes from that node to the other
nodes once it becomes available again so manual recovery can be necessary.

You should not rely on being able to do DDL during routine production
operation on BDR, or any other asynchronous multi-master replication setup.

**Comment by ringerc** [comment]
If you're looking for co-located HA-focused multi-master where you can assume good connectivity and low-latency, BDR isn't the product for you. For HA you're really better off sticking to streaming replication with active/standby failover.

**Comment by srkunze** [comment]
Thanks for your extensive answer, @ringerc . I understand that BDR design with different aspects in mind.

I am sure I am missing something but let me explain how my reasoning here is:

LXD containers make setting up BDR PostgreSQL instances easily. So, with little effort it's possible to have a new node joining an existing cluster and start replicating, serving, etc.

If you think that further, that means we would have an easy scale-up mechanism for PostgreSQL. So,
a) adding new nodes
b) removing superfluous nodes
c) disposing broken nodes
is possible with a simple LXD command. That given dynamic node allocation according to some usage stats would be possible as well. I would dare to call that cloud-ready or scale-out-ready.

> For HA you're really better off sticking to streaming replication with active/standby failover.

Maybe, I used the wrong term here but I get the feeling that cloudy scaling inherently comprises HA. From my perspective, active/standby failovers, however, seem more like a hand-crafted solution for specific usage scenarios.

**Comment by srkunze** [comment]
> BDR is intended for deployments where network connectivity may be
intermittent, slow and unreliable.

I understand that we can imagine some BDR setups where is this necessary. However, as described above one can also imagine other setups and given the infrastructure BDR provides in its current form, it's a tiny modification, that would enable even more scenarios (e.g. scale out).

> Under those circumstances, how should it know that you deleted the node? How long should it wait for it to reconnect before assuming it is permanently gone? There is no single right answer.

As long as the majority of the cluster is still available, then majority voting should fix all issues. Nodes joining in later, cannot vote anymore until there are up-to-date again.

> If you delete a node you have to tell BDR it isn't coming back.

It doesn't need to be perfect, but an option to disable this requirement would be great.

**Comment by ringerc** [comment]
On 24 November 2015 at 02:06, Sven R. Kunze <notifications@github.com>
wrote:

> BDR is intended for deployments where network connectivity may be
> intermittent, slow and unreliable.
>
> I understand that we can imagine some BDR setups where is this necessary.
> However, as described above one can also imagine other setups and given the
> infrastructure BDR provides in its current form, it's a tiny modification,
> that would enable even more scenarios (e.g. scale out).
>

It is not a tiny modification. It'd likely need a pretty significant set of
changes to make node join and removal lighter-weight processes and to make
BDR more tolerant of nodes parting unexpectedly.

PostgreSQL its self lacks support for multiple synchronous replicas,
anyway. It's being discussed for 9.6, but isn't possible in 9.4. You'd need
that to avoid data loss.

Perhaps you should look at Postgres-XC? It's probably closer to what you
are looking for.

Under those circumstances, how should it know that you deleted the node?
> How long should it wait for it to reconnect before assuming it is
> permanently gone? There is no single right answer.
>
> As long as the majority of the cluster is still available, then majority
> voting should fix all issues. Nodes joining in later, cannot vote anymore
> until there are up-to-date again.
>

You are hand-waving away hugely complex issues. It is really not that
simple. Majority voting does not fix issues that require consensus. Schema
changes require consensus.

If you're sure this is practical then I'd be interested in seeing a
prototype as a patch for BDR, to demonstrate how you'd achieve it. You will
have to demonstrate how it copes with a table structure change while some
nodes are offline/unreachable and there are changes for those tables queued
for replication in both directions. Expect to read a lot of source code.

I don't think this is a productive line of inquiry, but if you want to
spend your time on it I won't try to stop you. I don't promise to merge any
changes you make though - you'd have to be able to demonstrate that the
changes were robust, reliable and didn't break the use cases BDR is
primarily designed for.

> If you delete a node you have to tell BDR it isn't coming back.
>
> It doesn't need to be perfect, but an option to disable this requirement
> would be great.
>

I don't expect we'll add one. It'd be too easy for the whole system to
self-disassemble when faced with a network partition. Not to mention that
parting a node is a serious operation that requires manual recovery.

One thing I'd like to do is allow a parted node to still replay changes it
had queued to its ex-peers up to the point where it was parted, so even
once it's parted from the system if it's recovered it can replay changes.
That's still possibly a bit problematic because of conflict handling, but
nothing worse than an extended connectivity issue.

BDR also needs some major enhancements to handle unreachable node removal
better, without introducing inconsistencies. We need that anyway, though.

-- 
 Craig Ringer                   http://www.2ndQuadrant.com/
 PostgreSQL Development, 24x7 Support, Training & Services

**Comment by srkunze** [comment]
> It'd likely need a pretty significant set of changes to make node join and removal lighter-weight processes and to make BDR more tolerant of nodes parting unexpectedly.

I agree that you are more qualified to estimate the amount of changes required for this.

> PostgreSQL its self lacks support for multiple synchronous replicas, anyway. It's being discussed for 9.6, but isn't possible in 9.4.

You are quite a source of information. :)

> Perhaps you should look at Postgres-XC? It's probably closer to what you are looking for.

Thanks again for pointing me to an interesting alternative. I have to admit it looks quite promising. On a second  however, for my taste, there are too many moving and different parts that I would need to take care of.

Maybe, that was the major point why I loved BDR. All nodes are peers and can be replaced by another.

> Majority voting does not fix issues that require consensus. Schema changes require consensus.

I maybe then have not completely understood what consensus and voting means. Given my technical background in peer networks (crypto platforms, etc.), to me, voting is the process of reaching consensus among nodes. So, if there's a majority vote, all other nodes simply adhere to that.

> If you're sure this is practical then I'd be interested in seeing a prototype as a patch for BDR, to demonstrate how you'd achieve it.

You bet who's interested as well. However, I would rather use a drop-in solution that simply abstracts the peer and connection handling away. However, from what I can see such libraries are written with open non-controllable networks in mind which rather make everything quite complex and slow compared to a "all-peers-trust-each-other" scenario.

**Comment by srkunze** [comment]
> I don't expect we'll add one. It'd be too easy for the whole system to self-disassemble when faced with a network partition. Not to mention that parting a node is a serious operation that requires manual recovery.

Maybe, it's not that much of an issue if a cluster cannot be modified DDL-wise. However, I just stumbled over it and didn't expect it though.

Maybe, returning some kind of an error message would be great instead of waiting forever.

**Comment by ringerc** [comment]
> I maybe then have not completely understood what consensus and voting
means. Given my technical background in peer networks (crypto platforms,
etc.), to me, voting is the process of reaching consensus among nodes. So,
if there's a majority vote, all other nodes simply adhere to that.

"all other nodes simply adhere to that" isn't that simple. You're
transferring knowledge across domains without adjusting your assumptions. A
RDBMS has a lot of very complex state and state transitions that are quite
restricted. You can't just say "my peers told me to be like this, make it
so".

You want to ALTER the "users" table to add a new NOT NULL column
"recoverysecret".  Three nodes are reachable, one is unreachable. They vote
and agree that it's OK to do that, so each of them applies the schema
change.

Node 4 comes back online and says "I have some new users to tell you guys
about". The other 3 nodes respond with "those users are invalid, they are
missing the "recoverysecret" column". Node 4 is stuck. It can never replay
its changes to the other nodes.

The other nodes tell node 4 "now that you're back, let us tell you about
our new users too. But first, we changed the schema while you weren't
looking, here's the new schema." Node 4 says "um, that schema's nice and
all, but I have users here that don't have a value for the recoverysecret
column. What am I supposed to do now? The schema change cannot be applied
without a value for that column and I have no way to get one. So I can't
receive your new users either."

Node 4 is now stuck. It can't send or receive changes. You'll have to
remove it from the cluster then manually copy any new users it added while
offline to another node, adding a "recoverysecret" value while you're at it.

We could recover node 4 by discarding all pending changes and re-syncing
the table from an existing node. The operator will probably be annoyed by
the data loss, though, and the network cost of the table re-sync.

So... what? Do we disallow NOT NULL columns?

What about DROP COLUMN? We can ignore values for columns we don't have, but
we can't create the value when replicating to tables that didn't drop that
column.

What about foreign keys? There are endless complexities here, and you can't
handwave them away with "use majority voting".

Frankly, our current protocol is not sufficiently strict. What we need is a
2PC-based consensus protocol where we send the DDL to each node and say
"try to apply this and if you succeed PREPARE TRANSACTION so you can
promise to commit it successfully, then let me know". If all nodes succeed
we can COMMIT PREPARED; if one or more nodes fail we can ROLLBACK PREPARED.
This would handle differing local data and schema, which we can't currently
cope with properly.

Voting protocols work great with stable schemas and well defined data sets.
BDR does that for distributed sequences for that reason. They don't solve
everything.

Read the BDR code in detail. Prototype your ideas. Come back with a firm,
well defined description of how *exactly* you will handle these issues.
Keep in mind that it's asynchronous. Data can be committed on one node and
queued for replication, but not yet replicated to other nodes. The
replication progress between any pair of nodes is not the same as any other
pair of nodes, so a node can be "ahead" of another for one peer but
"behind" the same node for another peer. There is no one "now", each node's
concept of the current moment in time is separate.

I don't have time to go over this further given the current development
priorities. I'll be interested in whether you can design better handling
for these cases, but only if you come up with specific and definite ideas
based on a deep reading of the BDR sources and specific to how BDR,
PostgreSQL and logical decoding works. I don't plan to respond to further
general-discussion questions.

**Comment by ringerc** [comment]
On 25 November 2015 at 06:53, Sven R. Kunze <notifications@github.com>
wrote:

> I don't expect we'll add one. It'd be too easy for the whole system to
> self-disassemble when faced with a network partition. Not to mention that
> parting a node is a serious operation that requires manual recovery.
>
> Maybe, it's not that much of an issue if a cluster cannot be modified
> DDL-wise. However, I just stumbled over it and didn't expect it though.
>

The documentation in the development version discusses it in a bit more
detail.

http://bdr-project.org/docs/next/conflicts.html

http://bdr-project.org/docs/next/ddl-replication.html

We need to make it more explicit. Patches to further clarify the
documentation and add explanation where you would've expected to find it
would be greatly appreciated.

There are some significant improvements to DDL locking in 0.10.0 and
there's room for plenty more. What we have is quite heavy-handed and
rudimentary. With a few possible exceptions (like adding or dropping a
nullable column) it's not likely that we can ever avoid needing consensus
for DDL though.

Node join also needs consensus because we have to establish a consistent,
system-wide point in time where all replication slots start and where we're
caught up with our peers. We can't create a replication slot on an
absent/unreachable peer, and without that we cannot ensure that it will
retain data we need.


> Maybe, returning some kind of an error message would be great instead of
> waiting forever.
>
That'd break most applications, since they expect DDL to block until it can
run. In fact we do exactly that - return an error - when a distributed
sequence cannot acquire new values, and there have been no end of
complaints about it.

If you want an error, set a statement_timeout.

We could probably add a ddl_lock_wait_limit or something, so if the lock
isn't acquired after a certain number of seconds we abort the attempt. It'd
basically be statement_timeout with a different error message. Feel free to
send a patch for discussion.

The biggest improvement needed here is a way to make DDL lock status and
acquisition more visible. You can't see anything about DDL locking in the
pg_locks view, and that's frustrating for users. We need better visibility
there. Again, though, there are other priorities and there is not time to
develop it right now.

-- 
 Craig Ringer                   http://www.2ndQuadrant.com/
 PostgreSQL Development, 24x7 Support, Training & Services
